✨(feature) video permissions for the standalone site

qbey opened the pull request:
## Purpose

Permissions regarding videos' endpoint are made for LTI access, it needs to be updated to match future standalone site access.

## Proposal

- [x] Refactor the `IsVideoPlaylistAdmin` and `IsVideoOrganizationAdmin` to use `IsObjectPlaylist[...]` equivalent.
- [x] Allow all "specific actions" to playlist administrators and instructors
- [x] Add new permission tests 
    - [x] start_live
    - [x] stop_live
    - [x] harvest_live
    - [x] live_to_vod
    - [x] pairing_secret
    - [x] start_sharing
    - [x] navigate_sharing
    - [x] end_sharing
    - [x] participants_asking_to_join
    - [x] participants_in_discussion
    - [x] start_recording
    - [x] stop_recording
    - [x] stats
- [x] Add same permissions to the websocket part

lunika commented:
You should also check permissions in the websocket application. Websocket are used with video and they use the same permissions

kernicPanel reviewed (approved):
I feel like `src/backend/marsha/core/tests/api/video/test_start_stop_recording.py` should be split in two files, otherwise LGTM.

qbey commented:
> I feel like `src/backend/marsha/core/tests/api/video/test_start_stop_recording.py` should be split in two files, otherwise LGTM.

Yes, that's why I renamed the file `test_api_video_recording` into  `test_start_stop_recording` in a previous PR and also introduced the `ApiVideoStartRecordingTestCase` and `ApiVideoStopRecordingTestCase` here for a split to come, but I really don't want to include this in this PR.
The `test_shared_live_media` might also be split into one module per endpoint. But, as you know, those split are not the funniest part ^^

Thanks for the review!

lunika commented on src/backend/marsha/core/api/video.py:
```diff
@@ -76,14 +76,14 @@ def get_permissions(self):
             permission_classes = [
                 permissions.IsTokenResourceRouteObject
                 & (permissions.IsTokenInstructor | permissions.IsTokenAdmin)
-                | permissions.IsVideoPlaylistAdmin
```
In your commit body

`s/to have specifi permissions/to have specific permissions`

lunika commented on src/backend/marsha/core/api/video.py:
```diff
@@ -76,14 +76,14 @@ def get_permissions(self):
                 permissions.IsTokenResourceRouteObject
-                | permissions.IsVideoPlaylistAdmin
-                | permissions.IsVideoOrganizationAdmin
+                | permissions.IsObjectPlaylistAdmin
```
I think it can be `IsObjectPlaylistAdminOrInstructor`

lunika commented on src/backend/marsha/core/api/video.py:
```diff
@@ -76,14 +76,14 @@ def get_permissions(self):
                 & (permissions.IsTokenInstructor | permissions.IsTokenAdmin)
-                | permissions.IsVideoPlaylistAdmin
-                | permissions.IsVideoOrganizationAdmin
+                | permissions.IsObjectPlaylistAdmin
```
I think i can be `IsObjectPlaylistAdminOrInstructor`

lunika commented on src/backend/marsha/core/api/video.py:
```diff
@@ -98,7 +98,8 @@ def get_permissions(self):
         elif self.action in ["destroy"]:
             permission_classes = [
-                permissions.IsVideoPlaylistAdmin | permissions.IsVideoOrganizationAdmin
+                permissions.IsObjectPlaylistAdmin
```
I think it can be `IsObjectPlaylistAdminOrInstructor`

qbey replied:
Currently the playlist instructors cannot "create" video, should they be able to delete them ? or should we allow them to create video ? (note: the "create" permission has been added in another PR)

lunika replied:
then can do both. Create and delete.

lunika commented on src/backend/marsha/core/api/video.py:
```diff
@@ -206,8 +207,8 @@ def list(self, request, *args, **kwargs):
             | permissions.IsTokenResourceRouteObject & permissions.IsTokenAdmin
-            | permissions.IsVideoPlaylistAdmin
-            | permissions.IsVideoOrganizationAdmin
+            | permissions.IsObjectPlaylistAdmin
```
I think i can be `IsObjectPlaylistAdminOrInstructor`

lunika commented on src/backend/marsha/core/api/video.py:
```diff
@@ -121,10 +130,15 @@ def get_permissions(self):
                 permissions.IsTokenResourceRouteObject
                 & (permissions.IsTokenInstructor | permissions.IsTokenAdmin)
+                # With standalone site, playlist admin or instructor can access
+                | permissions.IsObjectPlaylistAdmin
```
I think it can be `IsObjectPlaylistAdminOrInstructor`

lunika commented on src/backend/marsha/core/tests/api/video/test_harvest_live.py:
```diff
@@ -22,6 +23,86 @@ class VideoHarvestLiveAPITest(TestCase):
+
+        self.assertEqual(response.status_code, 403)
+
+    @override_settings(LIVE_CHAT_ENABLED=True)
```
as you are only testing the live_state, you can remove this and everything related to the chat like the `close_room` mock and the `mock_jwt_encode`

lunika commented on src/backend/marsha/core/tests/api/video/test_live_participants_asking_to_join.py:
```diff
@@ -17,6 +27,183 @@ class VideoParticipantsAskingToJoinAPITest(TestCase):
+        """Assert the user can manage participants (POST and DELETE)."""
+        with mock.patch.object(
+            channel_layers_utils, "dispatch_video_to_groups"
+        ) as mock_dispatch_video_to_groups, mock.patch(
+            "marsha.core.serializers.xmpp_utils.generate_jwt"
+        ) as mock_jwt_encode:
+            mock_jwt_encode.return_value = "xmpp_jwt"
```
I don't think the xmpp part is needed

lunika commented on src/backend/marsha/core/tests/api/video/test_live_participants_joined.py:
```diff
@@ -17,6 +27,189 @@ class VideoParticipantsJoinedAPITest(TestCase):
+        """Assert the user can manage participants (POST and DELETE)."""
+        with mock.patch.object(
+            channel_layers_utils, "dispatch_video_to_groups"
+        ) as mock_dispatch_video_to_groups, mock.patch(
+            "marsha.core.serializers.xmpp_utils.generate_jwt"
+        ) as mock_jwt_encode:
+            mock_jwt_encode.return_value = "xmpp_jwt"
```
Same here, the xmpp part is not needed IMO

lunika commented on src/backend/marsha/websocket/consumers/video.py:
```diff
@@ -62,8 +91,23 @@ async def _check_video_exists(self):
+            & (
+                Q(
+                    playlist__user_accesses__user_id=user_id,
+                    playlist__user_accesses__role=ADMINISTRATOR,
```
Also `INSTRUCTOR`

lunika commented:
We miss something else. There are more objects related to video model and we have to manage permissions on them too. We need to adapt the permissions for `TimedTextTrack`, `SharedLiveMedia`, `Thumbnail`, `LiveSession`.
Without this, implementing video in the standalone site will be incomplete.
